Add Playback node package

MrTibbles opened the pull request:
- Adds new `playback-node` package for use in node.js environments only
- Adds `isValidKeyPair` to named exports on `playback-utils` package
- Adds `generateAppKey` test util method

See #65  for context
____

Initially set out to simply import the `Key` class from `playback-utils` to then call `setCrypto` method on the `Jose` instance, like so:
```js
const Jose = require('jose-jwe-jws')

import NodeWebCrypto from 'node-webcrypto-ossl'
import { Key } from '@jaak/playback-utils'

const webcrypto = new NodeWebCrypto()

Jose.setCrypto(webcrypto)

export default Key
```

However kept running into the same problem of the `crypto` API within `Jose` being undefined:
```bash
    TypeError: Cannot read property 'subtle' of undefined

      2971 |     if ( extractable === void 0 ) extractable = false;
      2972 | 
    > 2973 |   return Jose$1.crypto.subtle
```

I believe this is _something_ to do with the compiler not knowing which `Jose` instance/import to use and _somehow_ losing the ability to reference the methods on the module. 

So instead i have created a new `Key` class module local to `playback-node`, differing only in the support for generating key pairs, preferring importing keys. The `generate` class method was removed on the basis of us preferring app keys to be generated and registered via the Application dashboard. The `importKeys` method will change/proxy to the new `fromJWK` method being developed on #71 once completed. 

This has been tested with the example playback server.

lmars commented:
@MrTibbles 

> Initially set out to simply import the Key class from playback-utils to then call setCrypto method on the Jose instance ...

> However kept running into the same problem of the crypto API within Jose being undefined ...

> So instead i have created a new Key class module local to playback-node, differing only in the support for generating key pairs, preferring importing keys

I don't think it makes sense to duplicate the `Key` class because we couldn't get the `setCrypto` call to wire up correctly.

Could we not export `setCrypto` from `@jaak/playback-utils` (e.g. [here](https://github.com/jaakmusic/jaak.js/blob/master/packages/playback-utils/src/modules/key.js#L3)) so we know it's always being called on the same instance of `Jose`?

We should aim to have most of the functionality in environment-agnostic packages, the only thing `playback-node` needs to do is add the polyfill right?

MrTibbles commented:
@lmars thanks, good suggestion - hadnt considered that.
____
### Changes in 1aa6647: 
- Added `setCryptoProvider` method to `playback-utils`
- Added `importKeys` to `Key` class on `playback-utils`
- Added test case for importing keys to `playback-utils`
- Changed `playback-node` to export the `Key` module from playback-utils
- Add call to `setCryptoProvider` in playback-node, passing the node polyfill for web crypto API

lmars commented on packages/playback-node/src/modules/PlaybackServer.js:
```diff
@@ -0,0 +1,71 @@
+
+import { RegisterDeviceMutation } from '../graphql'
+
+class PlaybackServer {
```
This looks suspiciously like [`Playback.Application`](https://github.com/jaakmusic/jaak.js/blob/master/packages/playback/src/index.js#L13), is there a reason you've put this here in the `playback-node` package?

If your intention is to rename `Application` to `PlaybackServer`, I disagree with the name change, because this really isn't a server (what does it serve?). If anything, it's a client that makes requests to the Playback API?

In the docs, we specifically talk about registering devices to applications, so the following feels natural:

```js
const app = new Playback.Application(key)

await app.registerDevice(device)
```

Sure you can then use this within your web server, but you may also use it inside a CLI, a local electron app or a browser playground environment?

MrTibbles replied:
> This looks suspiciously like Playback.Application

It is uncanny ;)

My thinking was that an application is considered client side code, so therefore not intended for these purposes, so felt the name `PlaybackServer` was better suited. But happy to discuss.

lmars replied:
> an application is considered client side code

I think you spend too much time on the client side 😝 

I'd consider my app to be both the client side and the server side, it's all one thing.

This specific part of my app (the bit that requires an App Key and registers devices) most likely runs on a server, but as I said it may also run in a CLI I make to list my tracks and manually register devices in the console.

See how [Github Apps](https://developer.github.com/apps/) are described, it's a very similar setup.

MrTibbles replied:
> I think you spend too much time on the client side 

✊

In fairness it is more confusing including the word "server" than "application", as you say it does not _serve_ anything but will predominantly live on a server. I will change to `Application` - thanks

MrTibbles replied:
Resolved in 56b53ad

lmars commented on packages/playback-node/src/modules/Application.js:
```diff
@@ -0,0 +1,71 @@
+
+import { RegisterDeviceMutation } from '../graphql'
+
+class Application {
```
Can we not just leave this in the `@jaak/playback` package and just export it? I don't see the need to have this in a Node.js specific package?

MrTibbles replied:
I can see your point, i went with the basis that the vast majority of use cases for this class will be in a node.js environment. Firstly is that fair to say?
Ignoring the server discussion, [electron environments](https://electronjs.org/docs/tutorial/development-environment) are node envs, which will be the other main use case for the JS implementation of Playback packages:
> Electron development is essentially Node.js development.

Therefore i did not feel it appropriate to have this within a package intended for browser use.

lmars replied:
I just don't see the need to introduce the restriction, what problem does it solve?

MrTibbles replied:
It keeps the named exports concise, avoiding any confusion with the user of the library. If it is not to be used in that intended environment, it shouldnt be there.

lmars replied:
Ok, I'll make the decision easier to make. I want to build a "Playback playground" where you can paste in your app key, register devices and play tracks, all in the browser.

MrTibbles replied:
Ok, that is feasible. I will change.

MrTibbles replied:
Amended in 23a4c27

lmars commented on packages/playback-node/src/modules/importKeys.js:
```diff
@@ -0,0 +1,49 @@
+    )
+  }
+
+  const key = new ECKey(keyFile, 'pem')
```
Given we now use JWKs everywhere (i.e. we have `Key.fromJWK` and output JWKs in the JAAK app), do we need this?

MrTibbles replied:
We sure can, will make that change now, thanks

MrTibbles replied:
Amended in 2fdd169

lmars commented on test/setup.js:
```diff
@@ -26,3 +26,11 @@ global.verifyJWS = async (jws, key) => {
   return payload
 }
+
+global.generateAppKey = (extractable = false) => {
```
Can we not just use [`Key.generate()`](https://github.com/jaakmusic/jaak.js/blob/master/packages/playback-utils/src/modules/key.js#L16-L23)?

MrTibbles replied:
This is left over from when i implemented `Key` class within this package, rather than importing from playback-utils. I will remove this in favour of the importing approach. Thanks

MrTibbles commented on packages/playback-node/README.md:
````diff
@@ -0,0 +1,59 @@
+```
+
+## Usage
+First, create a `Key` instance with an absolute file path to an app key.
````
`fs. readFileSync` throws an error as it looks within the directory the package is in (node_modules), not the calling location.

lmars replied:
Are you sure? This seems to work for me with `key.json` in my current directory regardless of where I put the source file:

```js
const fs = require('fs')

const data = fs.readFileSync('key.json')

console.log('key:', JSON.parse(data))
```

MrTibbles replied:
```
    console.info packages/playback-node/dist/index.js:26
      { Error: ENOENT: no such file or directory, open 'test-app-key.json'
          at Object.openSync (fs.js:449:3)
          at Object.readFileSync (fs.js:349:35)
          at readFileSync (/Users/Fred/sites/jaak/jaak.js/packages/playback-node/dist/index.js:24:18)
          at Object.importKey (/Users/Fred/sites/jaak/jaak.js/packages/playback-node/__tests__/index.test.js:23:25)
          at Object.asyncJestTest (/Users/Fred/sites/jaak/jaak.js/node_modules/jest-jasmine2/build/jasmine_async.js:108:37)
          at resolve (/Users/Fred/sites/jaak/jaak.js/node_modules/jest-jasmine2/build/queue_runner.js:56:12)
          at new Promise (<anonymous>)
          at mapper (/Users/Fred/sites/jaak/jaak.js/node_modules/jest-jasmine2/build/queue_runner.js:43:19)
          at promise.then (/Users/Fred/sites/jaak/jaak.js/node_modules/jest-jasmine2/build/queue_runner.js:87:41)
          at processTicksAndRejections (internal/process/next_tick.js:81:5)
        errno: -2,
        syscall: 'open',
        code: 'ENOENT',
        path: 'test-app-key.json' }
```

lmars replied:
Is `test-app-key.json` in your working directory?

MrTibbles replied:
uh, no it was in the `test/` directory.

lmars replied:
Ok, so does it work with `test/test-app-key.json`?

MrTibbles replied:
But does that not show the instability of not using an absolute file path? Surely it just makes life easier?

lmars replied:
I don't get why we're telling the user that it _must_ be an absolute path.

If my key is in `test-app-key.json`, then why do I need to use an absolute path?

The following works just fine:

```js
const Playback = require('@jaak/playback-node')

Playback
  .importKey('test-app-key.json')
  .then(key => console.log('key:', key))
```

MrTibbles replied:
Removed explicit use of "absolute" paths and return passed file path in returned error; in b0ac62c

lmars replied:
The docs still say "with an absolute file path to an app key"?

MrTibbles replied:
Missed that, thanks - see fedf0a1

MrTibbles commented on packages/playback-node/__tests__/__fixtures__/fixtures.json:
```diff
@@ -0,0 +1,40 @@
+      }
+    }
+  },
+  "registerDeviceResponse": {
```
Resolved in 72fd2b7

MrTibbles commented on packages/playback-node/__tests__/index.test.js:
```diff
@@ -0,0 +1,37 @@
+const fetchMock = require('fetch-mock')
```
Resolved in 72fd2b7

MrTibbles commented on packages/playback-node/README.md:
````diff
@@ -0,0 +1,59 @@
+Then instantiate a new `Application` instance with the `Key` instance.
+
+```js
+import { Application, importKeys, Key } from '@jaak/playback-node'
````
Resolved in 72fd2b7

MrTibbles commented on packages/playback-node/README.md:
```diff
@@ -0,0 +1,59 @@
+const keyfilePath = '/Absolute/File/Path/to/jaak-app-key.pem'
+
+importKeys(keyfilePath).then(appKey => {
+  const playbackApp = new Application(appKey, { uri: apiEndpoint })
```
Resolved in 72fd2b7

MrTibbles commented on packages/playback-node/README.md:
````diff
@@ -0,0 +1,59 @@
+})
+```
+
+## Registering devices
````
Resolved in 72fd2b7

MrTibbles commented on packages/playback-node/README.md:
```diff
@@ -0,0 +1,59 @@
+# `@jaak/playback-node`
+
+> Simplified interactions with the JAAK Playback API for node JS environment
```
Resolved in 72fd2b7

MrTibbles commented on packages/playback-node/package.json:
```diff
@@ -0,0 +1,69 @@
+    "pretest": "npm run build:test",
+    "test": "jest",
+    "lint:staged": "lint-staged",
+    "prepublishOnly": "npm run build && npm run build:browser",
```
Resolved in 72fd2b7

MrTibbles commented on packages/playback-node/src/graphql/index.js:
```diff
@@ -0,0 +1,13 @@
+const RegisterDeviceMutation = `
+  mutation RegisterDevice($input: RegisterDeviceInput!) {
```
Resolved in 72fd2b7

MrTibbles commented on packages/playback-node/package.json:
```diff
@@ -0,0 +1,69 @@
+    "@jaak/playback": "^0.2.0",
+    "@jaak/playback-utils": "^0.2.0",
+    "node-webcrypto-ossl": "^1.0.39",
+    "fs-extra": "^7.0.1"
```
I believe it was left over from the just being lifted from example playback server, blindly brought across without consideration; thanks - resolved in 72fd2b7

MrTibbles commented on packages/playback-node/src/modules/importKey.js:
```diff
@@ -0,0 +1,26 @@
+import { isNonEmptyString, Key } from '@jaak/playback-utils'
+
+/**
+ * Import App keys from local file system
```
Resolved in 72fd2b7

lmars replied:
It still says "App key", and it should say "**a** local file".

MrTibbles replied:
Apologies, tired eyes last night, amended in 6522191

MrTibbles commented on packages/playback-node/src/modules/importKey.js:
```diff
@@ -0,0 +1,26 @@
+    keyFile = fs.readFileSync(keyFilePath, 'utf8')
+  } catch (error) {
+    throw new Error(
+      'Error loading app keys, visit your app dashboard to generate one.'
```
Resolved in 72fd2b7

lmars replied:
It still doesn't include the thrown error? And it also still says "app key".

MrTibbles replied:
Amended to suggested error message above, 6522191

lmars replied:
Sorry, so what I mean here is that the caught `error` probably gives a better indication of why there is an error, so we shouldn't just swallow it.

The most likely error will be ENOENT (i.e. "no such file"), but it could also be EISDIR (i.e. "the path is a directory") or something else that would be useful in figuring out what the problem is.

We should just pass the error along:

```
`Error loading key from ${keyFilePath}: ${error}`
```

MrTibbles replied:
Amended in fedf0a1

MrTibbles commented on packages/playback/__tests__/index.test.js:
```diff
@@ -130,12 +159,12 @@ describe('Playback.Application', () => {
-      const key = await Playback.Key.generate()
-      const app = new Playback.Application(key, { uri: fixtures.baseURL })
+      const key = await Key.generate()
+      const server = new Application(key, { uri: fixtures.baseURL })
```
Resolved in 72fd2b7

MrTibbles commented on packages/playback-node/src/index.js:
```diff
@@ -0,0 +1,13 @@
+setCryptoProvider(webcrypto)
+
+export { Application, importKey, Key }
+export default Application
```
Can do yeah, sounds sensible. I do plan to change the Playback exports to be Application, Device and Key, ie.
```js
export { Application, Device, Key }

export default {
  Application,
  Device,
  Key
}
```
